File: solutions/cpp/0117.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() {}

    Node(int _val, Node* _left, Node* _right, Node* _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        Node* node = root;
        Node* curr = new Node(NULL);
        Node* prev = curr;

        while (node) {
            while (node) {
                curr->next = node->left;
                if (curr->next) curr = curr->next;
                curr->next = node->right;
                if (curr->next) curr = curr->next;
                node = node->next;
            }
            node = prev->next;
            curr = prev;
        }

        return root;
    }
};

File: solutions/python3/0117.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, left, right, next):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""


class Solution:
    def connect(self, root: 'Node') -> 'Node':
        node = root
        curr = Node(None, None, None, None)
        prev = curr

        while node:
            while node:
                curr.next = node.left
                curr = curr.next or curr
                curr.next = node.right
                curr = curr.next or curr
                node = node.next
            node = prev.next
            curr = prev

        return root
